x/batching/keeper/abci.go

```diff
@@ -0,0 +1,149 @@
+	if err != nil {
+		return err
+	}
+	err = k.IncrementCurrentBatchNum(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) ConstructBatch(ctx sdk.Context) (types.Batch, error) {
+	curBatchNum, err := k.GetCurrentBatchNum(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+	dataLeaves, dataRootHex, err := k.ConstructDataResultTree(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+	valLeaves, valRootHex, err := k.ConstructValidatorTree(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+
+	return types.Batch{
+		BatchNumber:      curBatchNum,
+		BlockHeight:      ctx.BlockHeight(),
+		DataResultRoot:   dataRootHex,
+		ValidatorRoot:    valRootHex,
+		DataResultLeaves: dataLeaves,
+		ValidatorLeaves:  valLeaves,
+		BlockTime:        ctx.BlockTime(),
+	}, nil
+}
+
+// ConstructDataResultTree constructs a data result tree based on the
+// batching-ready data results returned from the core contract and
+// returns its leaves and its hex-encoded tree root.
+func (k Keeper) ConstructDataResultTree(ctx sdk.Context) ([][]byte, string, error) {
+	coreContract, err := k.wasmStorageKeeper.GetCoreContractAddr(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	// TODO: Deal with offset and limits. (#313)
+	queryRes, err := k.wasmViewKeeper.QuerySmart(ctx, coreContract, []byte(`{"get_data_results_by_status":{"status": "tallied", "offset": 0, "limit": 100}}`))
+	if err != nil {
+		return nil, "", err
+	}
+	if string(queryRes) == "[]" {
+		return nil, "", err
+	}
+
+	var dataResults []tallytypes.DataResult
+	err = json.Unmarshal(queryRes, &dataResults)
+	if err != nil {
+		return nil, "", err
+	}
+
+	leaves := make([][]byte, len(dataResults))
+	for _, res := range dataResults {
+		resHash, err := hex.DecodeString(res.ID)
+		if err != nil {
+			return nil, "", err
+		}
+		leaves = append(leaves, resHash)
+	}
+
+	curRoot := utils.RootFromEntries(leaves)
+	prevRoot, err := k.GetPreviousDataResultRoot(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	root := utils.SuperRootWithEntry(curRoot, prevRoot)
+
+	// TODO update data result status on contract
+
+	return leaves, hex.EncodeToString(root), nil
+}
+
+// ConstructValidatorTree constructs a validator tree based on the
+// validators in the active set and their registered public keys.
+// It returns the tree's entries (unhashed leaf contents) and hex-encoded
+// root.
+func (k Keeper) ConstructValidatorTree(ctx sdk.Context) ([][]byte, string, error) {
+	var entries [][]byte
+	err := k.stakingKeeper.IterateLastValidatorPowers(ctx, func(valAddr sdk.ValAddress, power int64) (stop bool) {
+		pubKey, err := k.pubKeyKeeper.GetValidatorKeyAtIndex(ctx, valAddr, utils.SEDAKeysIndexSecp256k1)
+		if err != nil {
+			if errors.Is(err, collections.ErrNotFound) {
+				return false
+			}
+			panic(err)
+		}
+
+		pkBytes := pubKey.Bytes()
+		entry := make([]byte, len(pkBytes)+8)
+
+		copy(entry[:len(pkBytes)], pkBytes)
+		binary.BigEndian.PutUint64(entry[len(pkBytes):], uint64(power))
```

Comment:
Weren't we going to use a uint16 or uint32 for a percentage based representation of the voting power?

Reply:
You're right. And I updated the code, but I need to think about possible overflow when casting to uint32..

Reply:
Yeah, worrying about the overflow would be good. You could `mod` by `u32::MAX` or compare it to u32 max. Though golang might have some way to deal with already built-in, and I'm not familiar enough with it. On overflows, we could either error or cap it.